source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+- The attribute `mkShell` of the `pkgs` set is a function that is passed an attribute set as argument.
+  Its return value is also the result of the outer function.
+- The attribute set passed to `mkShell` has the attributes `buildInputs` (set to a list with one element: the `cowsay` attribute from `pkgs`) and `shellHook` (set to an indented string).
+- The indented string contains an antiquotation, which will expand the value of `message` to yield `"hello world"`.
+
+
+Example:
+
+```nix
+{ config, pkgs, ... }: {
+
+  imports = [ ./hardware-configuration.nix ];
+
+  environment.systemPackages = with pkgs; [ git ];
+
+  # ...
+
+}
+```
+
+This example is (part of) a NixOS configuration.
+
+Explanation:
+
+- This expression is a function that takes an attribute set as an argument.
+  It returns an attribute set.
+- The argument must at least have the attributes `config` and `pkgs`, and may have more attributes.
+- The returned attribute set contains the attributes `imports` and `environment`.
+  `imports` is a list with one element: a path to a file next to this Nix file, called `hardware-configuration.nix`.
+
+  :::{note}
+  `imports` is not the impure built-in `import`, but a regular attribute name!
+  :::
+- `environment` is itself an attribute set with one attribute `systemPackages`, which will evaluate to a list with one element: the `git` attribute from the `pkgs` set.
+- The `config` argument is not used.
+
+(mkDerivation-example)=
+Example:
+
+```nix
+{ lib, stdenv }:
+
+stdenv.mkDerivation rec {
+
+  pname = "hello";
+
+  version = "2.12";
+
+  src = builtins.fetchTarball {
+    url = "mirror://gnu/${pname}/${pname}-${version}.tar.gz";
+    sha256 = "1ayhp9v4m4rdhjmnl2bq3cibrbqqkgjbl3s7yk2nhlh8vj3ay16g";
+  };
+
+  meta = with lib; {
+    license = licenses.gpl3Plus;
+  };
+
+}
+```
+
+This example is a (simplified) package declaration from Nixpkgs.
+
+Explanation:
+
+- This expression is a function that takes an attribute set which must have exactly the attributes `lib` and `stdenv`.
+- It returns the result of evaluating the function `mkDerivaion`, which is an attribute of `stdenv`, applied to a recursive set.
+- The recursive set passed to `mkDerivation` uses its own `pname` and `version` attributes in the argument to the built-in function `fetchTarball`.
+- The `meta` attribute is itself an attribute set, where the `license` attribute has the value that was assigned to the nested attribute `lib.licenses.gpl3Plus`.
+
+## References
+
+- [Nix manual: Nix language][manual-language] - Nix language reference
+- [Nix manual: Built-in Functions][nix-builtins] - Nix language built-in functions
+- [Nixpkgs manual: Functions reference][nixpkgs-functions] - Nixpkgs function library
+- [Nixpkgs manual: Fetchers][nixpkgs-fetchers] - Nixpkgs fetcher library
+
+## Next steps
+
+To get things done:
+
+- [](declarative-reproducible-envs) – create reproducible shell environments from a Nix file
```

Comment:
FYI `[](declarative-reproducible-envs)` renders as the empty string in the document.